File: frontend/src/app/core/services/auth.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, signal } from '@angular/core';
import { Router } from '@angular/router';
import { Observable, tap } from 'rxjs';

import { environment } from '../../../environments/environment';
import { AuthResponse, CurrentUserResponse, UserResponse, UserRole } from '../models/auth.model';

interface LoginRequest {
  usernameOrEmail: string;
  password: string;
}

interface SignupRequest {
  username: string;
  email: string;
  mobile: string;
  password: string;
}

interface CreateUserRequest {
  username: string;
  email: string;
  mobile: string;
  temporaryPassword: string;
  role: Exclude<UserRole, 'NORMAL_USER'>;
}

interface ChangePasswordRequest {
  currentPassword: string;
  newPassword: string;
}

@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private readonly baseUrl = `${environment.apiUrl}`;
  private readonly tokenKey = 'docuvra.auth.token';

  readonly currentUser = signal<CurrentUserResponse | null>(null);

  constructor(
    private readonly http: HttpClient,
    private readonly router: Router
  ) {
  }

  get token(): string | null {
    return window.localStorage.getItem(this.tokenKey);
  }

  login(request: LoginRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/auth/login`, request).pipe(
      tap(response => this.storeSession(response))
    );
  }

  signup(request: SignupRequest): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(`${this.baseUrl}/auth/signup`, request).pipe(
      tap(response => this.storeSession(response))
    );
  }

  loadCurrentUser(): Observable<CurrentUserResponse> {
    return this.http.get<CurrentUserResponse>(`${this.baseUrl}/auth/me`).pipe(
      tap(user => this.currentUser.set(user))
    );
  }

  changePassword(request: ChangePasswordRequest): Observable<CurrentUserResponse> {
    return this.http.post<CurrentUserResponse>(`${this.baseUrl}/auth/change-password`, request).pipe(
      tap(user => this.currentUser.set(user))
    );
  }

  listUsers(): Observable<UserResponse[]> {
    return this.http.get<UserResponse[]>(`${this.baseUrl}/users`);
  }

  listMentionableUsers(): Observable<UserResponse[]> {
    return this.http.get<UserResponse[]>(`${this.baseUrl}/users/mentionable`);
  }

  createUser(request: CreateUserRequest): Observable<UserResponse> {
    return this.http.post<UserResponse>(`${this.baseUrl}/users`, request);
  }

  logout(): void {
    window.localStorage.removeItem(this.tokenKey);
    this.currentUser.set(null);
    void this.router.navigate(['/login']);
  }

  private storeSession(response: AuthResponse): void {
    window.localStorage.setItem(this.tokenKey, response.token);
    this.currentUser.set(response.user);
  }
}

File: backend/src/main/java/com/docuvra/service/AuthService.java
package com.docuvra.service;

import com.docuvra.dto.AuthResponse;
import com.docuvra.dto.ChangePasswordRequest;
import com.docuvra.dto.CreateUserRequest;
import com.docuvra.dto.CurrentUserResponse;
import com.docuvra.dto.LoginRequest;
import com.docuvra.dto.RegisterNormalUserRequest;
import com.docuvra.dto.UserResponse;
import com.docuvra.entity.UserEntity;
import com.docuvra.enums.UserRole;
import com.docuvra.config.SecurityProperties;
import com.docuvra.exception.AuthException;
import com.docuvra.exception.ForbiddenException;
import com.docuvra.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Locale;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenService jwtTokenService;
    private final CurrentUserService currentUserService;
    private final SecurityProperties securityProperties;

    @Transactional
    public AuthResponse login(LoginRequest request) {
        UserEntity user = userRepository.findByUsernameIgnoreCase(normalize(request.usernameOrEmail()))
                .or(() -> userRepository.findByEmailIgnoreCaseAndRole(normalize(request.usernameOrEmail()), UserRole.NORMAL_USER))
                .or(() -> userRepository.findByEmailIgnoreCaseAndRole(normalize(request.usernameOrEmail()), UserRole.STAFF))
                .or(() -> userRepository.findByEmailIgnoreCaseAndRole(normalize(request.usernameOrEmail()), UserRole.SUPERVISOR))
                .filter(UserEntity::isActive)
                .orElseThrow(() -> new AuthException("Invalid username or password."));

        if (!passwordEncoder.matches(request.password(), user.getPasswordHash())) {
            throw new AuthException("Invalid username or password.");
        }

        user.setLastLoginAt(LocalDateTime.now());
        user.setUpdatedAt(LocalDateTime.now());
        return new AuthResponse(jwtTokenService.createToken(user), toCurrentUser(user));
    }

    @Transactional
    public AuthResponse signup(RegisterNormalUserRequest request) {
        ensureLoginEnabled("Signup is disabled while login is disabled.");
        UserEntity user = createUser(
                request.username(),
                request.email(),
                request.mobile(),
                request.password(),
                UserRole.NORMAL_USER,
                false
        );
        return new AuthResponse(jwtTokenService.createToken(user), toCurrentUser(user));
    }

    @Transactional(readOnly = true)
    public List<UserResponse> listUsers() {
        ensureLoginEnabled("User management is disabled while login is disabled.");
        currentUserService.requireRole(UserRole.SUPERVISOR);
        return userRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(this::toUserResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<UserResponse> listUsers(String search) {
        currentUserService.currentUserEntity();
        String query = normalize(search);
        return userRepository.findAllByOrderByCreatedAtDesc().stream()
                .filter(UserEntity::isActive)
                .filter(user -> query.isBlank()
                        || user.getUsername().toLowerCase(Locale.ROOT).contains(query)
                        || user.getEmail().toLowerCase(Locale.ROOT).contains(query))
                .map(this::toUserResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<UserResponse> listStaffUsers() {
        ensureLoginEnabled("Staff management is disabled while login is disabled.");
        currentUserService.requireRole(UserRole.SUPERVISOR);
        return userRepository.findAllByRoleAndActiveTrue(UserRole.STAFF).stream()
                .map(this::toUserResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<UserResponse> listMentionableUsers() {
        currentUserService.currentUserEntity();
        return userRepository.findAllByOrderByCreatedAtDesc().stream()
                .filter(UserEntity::isActive)
                .map(this::toUserResponse)
                .toList();
    }

    @Transactional
    public UserResponse createManagedUser(CreateUserRequest request) {
        ensureLoginEnabled("User management is disabled while login is disabled.");
        currentUserService.requireRole(UserRole.SUPERVISOR);
        if (request.role() == UserRole.NORMAL_USER) {
            throw new ForbiddenException("Supervisors can create only staff and supervisor users here. Normal users can use signup.");
        }
        return toUserResponse(createUser(
                request.username(),
                request.email(),
                request.mobile(),
                request.temporaryPassword(),
                request.role(),
                request.role() == UserRole.STAFF
        ));
    }

    @Transactional
    public CurrentUserResponse changePassword(ChangePasswordRequest request) {
        UserEntity user = currentUserService.currentUserEntity();
        if (!passwordEncoder.matches(request.currentPassword(), user.getPasswordHash())) {
            throw new AuthException("Current password is incorrect.");
        }
        user.setPasswordHash(passwordEncoder.encode(request.newPassword()));
        user.setForcePasswordChange(false);
        user.setUpdatedAt(LocalDateTime.now());
        return toCurrentUser(user);
    }

    private UserEntity createUser(
            String username,
            String email,
            String mobile,
            String password,
            UserRole role,
            boolean forcePasswordChange
    ) {
        String normalizedUsername = normalize(username);
        String normalizedEmail = normalize(email);
        String normalizedMobile = normalizeMobile(mobile);

        if (userRepository.existsByUsernameIgnoreCase(normalizedUsername)) {
            throw new AuthException("Username is already used. Please choose a different username.");
        }
        if (userRepository.existsByEmailIgnoreCaseAndRole(normalizedEmail, role)) {
            throw new AuthException("Email is already used for this role.");
        }
        if (userRepository.existsByMobileAndRole(normalizedMobile, role)) {
            throw new AuthException("Mobile number is already used for this role.");
        }

        LocalDateTime now = LocalDateTime.now();
        UserEntity user = new UserEntity();
        user.setId(UUID.randomUUID());
        user.setUsername(normalizedUsername);
        user.setEmail(normalizedEmail);
        user.setMobile(normalizedMobile);
        user.setPasswordHash(passwordEncoder.encode(password));
        user.setRole(role);
        user.setActive(true);
        user.setForcePasswordChange(forcePasswordChange);
        user.setCreatedAt(now);
        user.setUpdatedAt(now);
        return userRepository.save(user);
    }

    private CurrentUserResponse toCurrentUser(UserEntity user) {
        return new CurrentUserResponse(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getMobile(),
                user.getRole(),
                user.isForcePasswordChange(),
                true
        );
    }

    private UserResponse toUserResponse(UserEntity user) {
        return new UserResponse(
                user.getId(),
                user.getUsername(),
                user.getEmail(),
                user.getMobile(),
                user.getRole(),
                user.isActive(),
                user.isForcePasswordChange(),
                user.getCreatedAt(),
                user.getLastLoginAt()
        );
    }

    private String normalize(String value) {
        return value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
    }

    private String normalizeMobile(String value) {
        return value == null ? "" : value.replaceAll("\\s+", "").trim();
    }

    private void ensureLoginEnabled(String message) {
        if (!securityProperties.loginEnabled()) {
            throw new ForbiddenException(message);
        }
    }
}
